mod/badges/set/abstract/02_badge_hierarchy.rb

```diff
@@ -0,0 +1,119 @@
+#! no set module
+
+# A BadgeHierarchy manages BadgeSets for one badge cardtype, for example
+# all badges related to metric answers
+module BadgeHierarchy
+  BADGE_TYPES =
+    [:metric, :project, :metric_value, :source, :wikirate_company].freeze
+
+  def self.for_type type_code
+    Card::Set::Type.const_get("#{type_code.to_s.camelcase}::BadgeHierarchy")
+  end
+
+  attr_reader :badge_level, :levels, :levels_descending, :badge_action
+
+  def add_badge_set action, badge_set, &count_wql
+    @map ||= {}
+    @map[action] = Abstract::BadgeSet.new badge_set, &count_wql
+  end
+
+  def add_affinity_badge_set action, map
+    @map ||= {}
+    @map[action] = {}
+    map.each do |affinity, affinity_badge_set|
+      @map[action][affinity] = Abstract::BadgeSet.new affinity_badge_set
+    end
+  end
+
+  def create_type_count type_id
+    lambda do |user_id|
+      {
+        type_id: type_id,
+        created_by: user_id
+      }
+    end
+  end
+
+  def type_plus_right_count type_id, right_id, relation_to_user
+    lambda do |user_id|
+      {
+        left: { type_id: type_id },
+        right_id: right_id,
+        relation_to_user => user_id
+      }
+    end
+  end
+
+  def type_plus_right_edited_count type_id, right_id
+    type_plus_right_count(type_id, right_id, :edited_by)
+  end
+
+  def vote_count type_id
+    lambda do |user_id|
+      user = user_id ? Card[user_id] : Auth.current
+      vote_card_ids = [user.upvotes_card.id, user.downvotes_card.id].compact
+      next nil unless vote_card_ids.present?
+      {
+        type_id: type_id,
+        referred_to_by: { id: ["in"] + vote_card_ids }
+      }
+    end
+  end
```

Comment:
probably possible to reuse the counts of the tabs on the profile page but not worth the effort since we are going to introduce a new caching system there